pkg/audit/manager.go

```diff
@@ -626,3 +645,31 @@ func logViolation(l logr.Logger, constraint *unstructured.Unstructured, enforcem
 		logging.ResourceName, violation.rname,
 	)
 }
+
+func emitEvent(constraint *unstructured.Unstructured, enforcementAction string, violation auditResult, eventRecorder record.EventRecorder, timestamp string, gkNamespace string) {
+	annotations := map[string]string{
+		"process":                   "audit",
+		"auditTimestamp":            timestamp,
+		logging.EventType:           "violation_audited",
+		logging.ConstraintKind:      constraint.GetKind(),
+		logging.ConstraintName:      constraint.GetName(),
+		logging.ConstraintNamespace: constraint.GetNamespace(),
+		logging.ConstraintAction:    enforcementAction,
+		logging.ResourceKind:        violation.rkind,
+		logging.ResourceNamespace:   violation.rnamespace,
+		logging.ResourceName:        violation.rname,
+	}
+	reason := "AuditViolation"
+	ref := getViolationRef(gkNamespace, violation.rkind, violation.rname, violation.rnamespace, constraint.GetKind(), constraint.GetName(), constraint.GetNamespace())
+
+	eventRecorder.AnnotatedEventf(ref, annotations, corev1.EventTypeWarning, reason, "Timestamp: %s, Resource Namespace: %s, Constraint: %s, Message: %s", timestamp, violation.rnamespace, constraint.GetName(), violation.message)
+}
+
+func getViolationRef(gkNamespace, rkind, rname, rnamespace, ckind, cname, cnamespace string) *corev1.ObjectReference {
+	return &corev1.ObjectReference{
+		Kind:      rkind,
+		Name:      rname,
+		UID:       types.UID(rkind + rname + rnamespace + ckind + cname + cnamespace),
```

Comment:
We should use the full unique identifier (group, kind, namespace, name) for both the constraint and the kind.

We should also use a separator character to prevent accidental overlap between two tokens causing redundancy.

Example:

"OneMore" + "Thing" == "One" + "MoreThing"

The separator character should be disallowed for use in any of the tokens... I think `/` fits the bill... does it work in UID? Any other character that is disallowed by all tokens would work.

Reply:
We don't actually keep track of violating resource's group today, only kind, name, and namespace. To be consistent, do you think this is specific to emitting events or this information should also be part of the constraint status and logs?

Reply:
This should probably be part of everything, as without a group, the Kind may be ambiguous.

Good catch if we are not using it elsewhere.

Reply:
@maxsmythe should we add `group` as a separate PR since this changes the schema of logs and constraint status and it should be addressed separately.

Reply:
That works for me. I'm surprised this is the first time we're noticing that we should add group.

Reply:
Created #748